src/opensearch.py

```diff
@@ -0,0 +1,187 @@
+            logger.error(f"Failed to install opensearch. \n{e}")
+            raise OpenSearchInstallError()
+
+        plugs = [
+            "log-observe",
+            "mount-observe",
+            "process-control",
+            "procsys-read",
+            "system-observe",
+        ]
+        for plug in plugs:
+            self._run_cmd(f"snap connect opensearch:{plug}")
+
+    def start(self):
+        """Start the snap exposed "daemon" service."""
+        if not self._opensearch.present:
+            raise OpenSearchMissingError()
+
+        if self._opensearch.services[self.SERVICE_NAME]["active"]:
+            logger.info(
+                f"Not doing anything, the opensearch.{self.SERVICE_NAME} service is already started."
+            )
+            return
+
+        try:
+            self._opensearch.start([self.SERVICE_NAME])
+        except SnapError as e:
+            logger.error(f"Failed to start the opensearch.{self.SERVICE_NAME} service. \n{e}")
+            raise OpenSearchStartError()
+
+    def stop(self):
+        """Stop the snap exposed "daemon" service."""
+        if not self._opensearch.present:
+            raise OpenSearchMissingError()
+
+        try:
+            self._opensearch.stop([self.SERVICE_NAME])
+        except SnapError as e:
+            logger.error(f"Failed to stop the opensearch.{self.SERVICE_NAME} service. \n{e}")
+            raise OpenSearchStopError()
+
+    def restart(self):
+        """Restart the snap exposed "daemon" service."""
+        if not self._opensearch.present:
+            raise OpenSearchMissingError()
+
+        try:
+            self._opensearch.restart([self.SERVICE_NAME])
+        except SnapError as e:
+            logger.error(f"Failed to restart the opensearch.{self.SERVICE_NAME} service. \n{e}")
+            raise OpenSearchRestartError()
+
+    def _build_paths(self) -> Paths:
+        """Builds a Path object.
+
+        The main paths are:
+          - OPENSEARCH_HOME: read-only path ($SNAP/..), where the opensearch binaries are
+          - OPENSEARCH_CONF: writeable by root or snap_daemon ($SNAP_COMMON) where config files are
+        """
+        return Paths(
+            home="/var/snap/opensearch/current",
+            conf="/var/snap/opensearch/common/config",
+            data="/var/snap/opensearch/common/data",
+            logs="/var/snap/opensearch/common/logs",
+            jdk="/var/snap/opensearch/current/jdk",
+            tmp="/var/snap/opensearch/common/tmp",
+        )
+
+
+class OpenSearchTarball(OpenSearchDistribution):
+    """Snap distribution of opensearch, only overrides properties and logic proper to the snap."""
+
+    def __init__(self, charm, peer_relation: str):
+        super().__init__(charm, peer_relation)
+
+    def install(self):
+        """Download and Un-tar the opensearch distro."""
+        response = requests.get(
+            "https://artifacts.opensearch.org/releases/bundle/opensearch/2.3.0/opensearch-2.3.0-linux-x64.tar.gz"
```

Comment:
Is this the best place to get opensearch? Should we be building and storing it ourselves? 

Also, maybe parameterise the opensearch version so it can be easily updated and accessed from elsewhere.

Reply:
This is only a temporary way of installing OpenSearch while we wait for the snap to be published. This class will be replaced by `OpenSearchSnap` once the snap is published.